separate pagination for versions

alemela opened the pull request:
Testing some ways to visualize versions history I decided to add a small header and a pagination with external links, because I think having everything in the same page may be confusing. Instead, if you want all in one page we can transform external links to anchors.

retog commented:
@alemela, I'm a bit surprised you couldn't use more from https://github.com/zazukoians/municipality-browser. I think seeing the current name of the municipality is a useful feature and "navigating the versions" for me implies one can easily go from one version to another as this is possible when the versions are shown inline now to navigate to the next version one has to first find their way back.

Two more technical consideration:

- I think in the <catchall> matcher there shouldn’t be anything site or ontology specific, this should be a fallback if there is no specific template, I can imagine creating a *generic* mechanism to tell the fallback template that a specific set of triples/properties should be ignored. So that a more specific template could let the fallback template know that say, dc:title has already been rendered and need no longer show up in the table.
- The idea is that everything site specific goes into site-matchers.ttl so that the default matchers.ttl can be used (don't remember why we started overwriting matchers.ttl in this project)

alemela commented:
With the new commit:

1. all the new matchers are in the site-matcher
2. now you don't leave the municipality page to see versions and you have a pagination to move between them

But I think you shouldn't merge until we've remove the timeout at line 284. I don't know how to wait until fetch are completed; any callback?

retog commented:
Hi @alemela, the timeout is no longer needed with the latest ld2h version

alemela commented:
Hi @retog, thanks! I've update the code removing the timeout.

retog commented:
Great progress! The main remaining issue is the header and footer. Another issue is that one can't collapse the version again, the easiest way to fix this is to have the latest version expanded by default.

This is my feedback from trying it out, I'll have a quick look at the code now.

alemela commented:
I agree with the collapse feature, I'll add!

Any hint for not showing header and footer? The only idea I had is to hide them via jQuery...

retog commented on data/public/rdf2h/site-matchers.ttl:
```diff
@@ -233,3 +237,111 @@
+        r2h:context r2h:Default;
+        r2h:mustache '''
+        {{{:continue}}}
+        <script href="https://cdnjs.cloudflare.com/ajax/libs/jquery/2.2.4/jquery.min.js"></script>      
```
jquery 2.1.4 is loaded in in the base template, couldn't we just update the version there?

retog commented on data/public/rdf2h/site-matchers.ttl:
```diff
@@ -281,14 +281,14 @@
             });
           }
-          setTimeout(function() {
+          $( document ).ready(function() {
```
I actually meant that the promise returned by LD2h.expand() is no longer satisfied too early. The respective bug was fixed with LD2h 0.4.0, I've just now updated trifid-ld to use 0.4.1 instead of 0.3.0.

retog commented on data/public/rdf2h/site-matchers.ttl:
```diff
@@ -281,14 +281,14 @@
             $("#history").insertBefore( $( "#footer" ) );
             $("#history").show();
-          }, 5000);
+          });
```
Well, I don't see why document.ready should work here, isn't that the same as the jQuery shortcut here: https://github.com/zazukoians/trifid-ld/blob/master/data/templates/graph.html#L20?

I think the easiest would be to assign the promise returned on on https://github.com/zazukoians/trifid-ld/blob/master/data/templates/graph.html#L27 to a global variable (say `ld2hCompleted`) so that here you can call `ld2hCompleted.then(function()`.

retog commented:
@alemela easier than adding the collapse feature is to just show the latest version (the need for collapsing only comes from seeing things collapsed at the beginning, if they are never "collapsed" that's OK too).

retog commented:
It would be good if `matchers.ttl`could now be removed from the project, so that the default trifid-ld `matchers.ttl` is used and all data.admin specific stuff is in `site-matchers.ttl`

retog commented:
> Any hint for not showing header and footer? The only idea I had is to hide them via jQuery...

using different contexts.

alemela commented:
@retog I'll use the ld2h promise but it requires a cross repo PR so I'll do as last thing. Trying to use only site-matchers and removing header and footer for version I find out the same problem. How can I "disable" one default template?

I can't figure out how to use different context because it's a default context template and I want to use catchall...

retog commented:
@alemela thanks for the question about disabling an inherited template :)
 
What you need to do is to create a context switching loophole (csl), assuming you want to "disable" a matcher m for  the context r2h:Default, what you do is create a loophole portal matcher p so that p r2h:before m, the template of the portal does nothing but switching the mode, i.e. {{{:continue myns:myLoopholeMode}}} (note that here you could user ":render ." instead of :continue but this would be less efficient). Now the only missing part is the exit of the loophole, for that we create a matcher e so that m r2h:before e which switches back the mode, i.e. {{{:continue r2h:Default}}} (here you have to use :continue).

alemela commented:
@retog thanks very much! With your explanation I've removed the unwanted templates. I had to commit again the matchers.ttl (but the trifid one) because to refer to its template I had to add a `@base` triple at the beginning. That's the only difference to add in trifid.

How do you want to conclude this PR? The only remaining part is the promise but to provide you a working PR I must edit also trifid repo...

retog commented:
@alemela, can you make a PR against Trifid? thx

retog commented on data/public/rdf2h/site-matchers.ttl:
```diff
@@ -337,14 +370,31 @@ Linked Data Pilot</p></div>
     r2h:before r2hp:specific;
     r2h:before <page>;
     r2h:before <headline>;
+    r2h:before <endEscapeDadminCaps>;
```
It seems this should be before something (the skipped matcher) that is before  endEscapeDadminCaps, this doesn't make this statement false but superfluos. I guess all matchers in matchers.ttl should have absolute URIs so that they can be referenced here.

retog commented:
I'm taking care of https://github.com/zazukoians/trifid-ld/issues/35

alemela commented:
I've removed the matchers.ttl here and added references but trying locally with docker the trifid matchers.ttl isn't the last you updated, but the old one. Don't know why :S.

The "endEscapeDadminCaps" matcher is the closing template for avoid header/footer in the version snapshot (we avoid <capsdadmin> and the trifid's <main>). I change context before writing the version context, I avoid those 2 matchers, then I close the loop with "endEscapeDadminCaps". With you it does not have the same behavior?

retog commented:
Did you try `docker build --no-cache .` ?

retog commented on data/public/rdf2h/site-matchers.ttl:
```diff
@@ -1,4 +1,4 @@
-@base <http://example.com/> .
+@base <http://data.admin.ch/> .
```
http://data.admin.ch/rdf2h/matchers/

retog commented on data/public/rdf2h/site-matchers.ttl:
```diff
@@ -233,3 +252,162 @@
+        {{{:continue mode:loopholeTrifidCaps}}}
+        '''
+    ];
+    r2h:before <caps>.
```
there is no <caps> this should be 
    r2h:before <http://zazukoians.org/2016/generic-rdf2h-matchers/caps>

retog commented on data/public/rdf2h/site-matchers.ttl:
```diff
@@ -8,8 +8,27 @@
+            Code on <a href="https://github.com/zazuko/data.admin.ch" style="color:white"><b>github</b></a>
+        </div>'''
+  ];
+  r2h:before <catchall>.
```
<catchall> is not in current namespace

retog commented on data/public/rdf2h/site-matchers.ttl:
```diff
@@ -233,3 +252,162 @@
+    r2h:before <endEscapeDadminCaps>;
+].
+
+<capsdadmin> r2h:before <endEscapeDadminCaps>.
```
Not seeing the rationale for this, I thing 332 states the only relevant condition

retog commented on data/public/rdf2h/site-matchers.ttl:
```diff
@@ -233,3 +252,162 @@
+].
+
+<capsdadmin> r2h:before <endEscapeDadminCaps>.
+<http://zazukoians.org/2016/generic-rdf2h-matchers/main> r2h:before <endEscapeDadminCaps>.
```
Again, don't see the rationale, for the loophole to work the corrected statement of line 330 and 332 should be enough.

alemela commented:
Statements of line 330 - 332 remove the trifid header/footer leaving the dadmin one, and it's ok. But when we fetch versions we restart the whole matchers loop and again it puts header/footer (and "main") that we want for the whole page but not for the versions. So that why I've written two different loop:

- we don't want nevere the trifid H/F (loop #1)
- when version we don't also want the dadmin H/F (loop #2)

alemela commented:
Seemed quite weird but were just two logical contexts mixed together. So I write down the context routing: `r2h:Website` if `r2h:object schema:WebPage` and never return to `r2h:Default`. Instead, if no website it enter inside `capsdadmin` that stands upon the standard matcher.

retog commented:
Please merge down master occasionally.

Must have today:
 - The latest not the oldest incarnation should be shown by default
-  The name of the (oldest incarnation of the municipality should show up on top)

Other things:

- Create a specifc matcher for things with dbpedia summary and logo, the details from dbpedia should be collapsed

![data-admin-ch_mashup](https://cloud.githubusercontent.com/assets/110756/16224309/bfb6aa0a-37a1-11e6-9724-5671c330c5f7.png)

- White background: main data
- owl:same Targets from admin.ch domain: light greenish background
- other owl:same Targets (e.g. DBpedia): reddish/pinkish background

alemela commented:
I've completed the must have. Could you be a little more specific about the next feature request? Thanks!

retog commented:
Thanks @alemela. I'll open issues with more detailed description of the feature requests.

retog commented:
I've created https://github.com/zazuko/data.admin.ch/issues/15. Also I added you as collaborator, so you can create branches on this project and make the pull-request from there, please never commit to master (but make PRs).
